File: python/exer01c_max_div.py
'''
MAXIMUM NUMBER OF DIVISORS
'''

import time
import threading



lk = threading.Lock()
final_res = { 'value': 0, 'numdiv': 0 }



def prepare_arg(rng_start: int, rng_end: int, num_threads: int) -> list[dict]:
    rng_block = (rng_end - rng_start + 1) // num_threads
    rng_a = rng_start
    lst_arg = []

    for i in range(num_threads):
        rng_b = rng_a + rng_block - 1 if i < num_threads - 1 else rng_end
        lst_arg.append({ 'start': rng_a, 'end': rng_b })
        rng_a += rng_block

    return lst_arg



def do_task(arg: dict):
    res_value = 0
    res_numdiv = 0

    for i in range(arg['start'], arg['end'] + 1):
        numdiv = 0

        for j in range(1, i // 2):
            if i % j == 0:
                numdiv += 1

        if res_numdiv < numdiv:
            res_numdiv = numdiv
            res_value = i

    with lk:
        if final_res['numdiv'] < res_numdiv:
            final_res['numdiv'] = res_numdiv
            final_res['value'] = res_value



##########################################################



RANGE_START = 1
RANGE_END = 30000
NUM_THREADS = 8

lst_worker_arg = prepare_arg(RANGE_START, RANGE_END, NUM_THREADS)

lstth = [
    threading.Thread( target=do_task, args=(arg,) )
    for arg in lst_worker_arg
]


tp_start = time.time()

for th in lstth:
    th.start()

for th in lstth:
    th.join()

time_elapsed = time.time() - tp_start


print('The integer which has largest number of divisors is', final_res['value'])
print('The largest number of divisor is', final_res['numdiv'])
print('Time elapsed =', time_elapsed)

File: js-nodejs/exer01c-max-div.js
/*
MAXIMUM NUMBER OF DIVISORS
*/

import * as mylib from './mylib.js';
import { Mutex } from './mylib_mutex.js';
import { isMainThread, workerData, parentPort } from 'worker_threads';


//---------------------------------------------
//            WORKER THREAD SECTION
//---------------------------------------------


const doTask = arg => {
  let resValue = 0;
  let resNumDiv = 0;

  for (let i = arg.iStart; i <= arg.iEnd; ++i) {
    let numDiv = 0;

    for (let j = i / 2; j > 0; --j)
        if (i % j == 0)
            ++numDiv;

    if (resNumDiv < numDiv) {
        resNumDiv = numDiv;
        resValue = i;
    }
  }

  parentPort.postMessage({ resValue, resNumDiv });
};

const workerFunc = () => {
  try {
    const arg = workerData;
    doTask(arg);
  } catch (error) {
    console.error(error);
  }
};


//---------------------------------------------
//             MAIN THREAD SECTION
//---------------------------------------------


const prepareArg = (rangeStart, rangeEnd, numThreads) => {
  let rangeA, rangeB, rangeBlock;

  rangeBlock = (rangeEnd - rangeStart + 1) / numThreads;
  rangeA = rangeStart;

  let lstWorkerArg = [];

  for (let i = 0; i < numThreads; ++i, rangeA += rangeBlock) {
      rangeB = rangeA + rangeBlock - 1;

      if (i == numThreads - 1)
          rangeB = rangeEnd;

      lstWorkerArg.push({ iStart: rangeA, iEnd: rangeB });
  }

  return lstWorkerArg;
};


const updateFinalResult = (mutex, finalRes, value, numDiv) => {
  // Should protect by a mutex?
  mutex.runExclusive(() => {
    if (finalRes.numDiv < numDiv) {
      finalRes.numDiv = numDiv;
      finalRes.value = value;
    }
  });
};


const mainFunc = async () => {
  try {

    const RANGE_START = 1;
    const RANGE_END = 100000;
    const NUM_THREADS = 8;

    const lstWorkerArg = prepareArg(RANGE_START, RANGE_END, NUM_THREADS);
    const finalRes = { value: 0, numDiv: 0 };

    const lstTh = new Array(NUM_THREADS);
    const mutex = new Mutex();

    const tpStart = process.hrtime();

    for (let i = 0; i < NUM_THREADS; ++i) {
      const arg = lstWorkerArg[i];
      const [worker, prom] = mylib.createThread(new URL(import.meta.url), arg);
      worker.on('message', res => updateFinalResult(mutex, finalRes, ...Object.values(res)));
      lstTh[i] = [worker, prom];
    }

    await Promise.all(lstTh.map(([_,pr]) => pr));
    const timeElapsed = mylib.hrtimeToNumber(process.hrtime(tpStart));

    console.log('The integer which has largest number of divisors is', finalRes.value);
    console.log('The largest number of divisor is', finalRes.numDiv);
    console.log('Time elapsed =', timeElapsed);

  } catch (error) {
    console.error(error);
  }
};


if (isMainThread) {
  await mainFunc();
} else {
  workerFunc();
}
